### DISCUSSION TITLE
[QUESTION] Calculations regarding calculate_per_token_loss parameter

### DISCUSSION THREAD

**Discussion by clarence-lee-sheng** [NVIDIA/Megatron-LM#1100] [Q&A]
In line 231-233 in megatron/core/pipeline_parallel/schedules.py ([megatron/core/pipeline_parallel/schedules.py](https://github.com/NVIDIA/Megatron-LM/blob/80c7c6e936e6868fe251eabb079ae5d84ee28311/megatron/core/pipeline_parallel/schedules.py#L231)), I have two questions: 
1. Why are we dividing by num_tokens when the conditional is "if not config.calculate_per_token_loss"
2. What is the purpose of dividing by num_microbatches if it is a constant, and if it is important, why do we not also divide by num_microbatches outside of the condition for the config.calculate_per_token_loss true case.

**Comment by Advait251206** [comment]
The key point is that `calculate_per_token_loss` changes **how the loss is normalized**, and the two code paths are intended to produce gradients with different scaling semantics.

Regarding your questions:

### 1. Why divide by `num_tokens` when `calculate_per_token_loss` is **False**?

At first glance, this does seem counterintuitive based on the flag name. The important thing to verify is what `num_tokens` actually represents at that point in the pipeline.

In many Megatron-LM loss implementations:

- `num_tokens` is the number of valid (non-masked) tokens.
- The loss returned by the model may already be either summed or averaged before it reaches the pipeline schedule.

If the upstream loss function already performs part of the normalization, then this division may simply be completing the desired scaling for gradient accumulation rather than switching to a "per-token" objective.

Without considering the surrounding code (particularly where the loss is produced), it's difficult to conclude that this is incorrect—it may simply reflect where Megatron-LM chooses to apply the normalization.

### 2. Why divide by `num_microbatches`?

During gradient accumulation, the effective batch consists of multiple microbatches:

```text
Global batch
    ↓
Microbatch 1
Microbatch 2
...
Microbatch N
```

If each microbatch contributes its full loss to the gradients, the accumulated gradients would be approximately `N` times larger than those from a single large batch.

Dividing by `num_microbatches` ensures that the accumulated gradient corresponds to the average over the effective batch, making training independent of the number of microbatches used for gradient accumulation.

### 3. Why isn't `num_microbatches` also applied when `calculate_per_token_loss=True`?

This is the part that deserves closer inspection.

One possible explanation is that when `calculate_per_token_loss=True`, the loss returned from the model is already normalized across all valid tokens (potentially accounting for gradient accumulation elsewhere), so an additional division by `num_microbatches` would result in double normalization.

Another possibility is that the normalization is deferred until after aggregating token counts across all microbatches, allowing the final loss to be computed as:

```text
total_loss / total_valid_tokens
```

instead of:

```text
(loss₁ / tokens₁ + loss₂ / tokens₂ + ...)
```

which are generally **not** equivalent when microbatches contain different numbers of valid tokens.

So the asymmetry is likely intentional, assuming the two branches receive differently normalized loss values. If both branches receive identically scaled losses, then your observation would indeed suggest an inconsistency worth investigating.

Overall, your questions are valid, and the naming of `calculate_per_token_loss` can make this section of the code difficult to follow. The definitive answer depends on how the loss tensor is produced before it reaches `pipeline_parallel/schedules.py`, but the differing normalization strategies are likely intended to support two distinct loss-scaling modes rather than being an oversight.

---

If this answer helped or pointed you in the right direction, I'd appreciate it if you could mark it as the accepted answer so it's easier for others with the same issue to find.

Also, if you found my contribution useful, I'd appreciate it if you could check out my GitHub profile, follow me, and star any repositories you find interesting.

GitHub: https://github.com/Advait251206
